aten/src/ATen/native/cuda/Indexing.cu

```diff
@@ -840,6 +842,90 @@ Tensor index_select_cuda(const Tensor& self, int64_t dim, const Tensor& index) {
   return out;
 }
 
+template<typename T>
+struct NonZeroOp
+{
+    __host__ __device__ __forceinline__ bool operator()(const T& a) const {
+      return (a!=T(0));
+    }
+};
+
+template<typename scalar_t>
+void nonzero_cuda_out_impl(const Tensor& self, Tensor& out){
+  Tensor self_ = self.contiguous();
+  int N = self_.numel();
+  //compute number of nonzero elements
+  void * d_temp_storage=NULL;
+  size_t temp_storage_bytes=0;
+  auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+  auto num_nonzeros = allocator.allocate(sizeof(int));
+  cub::TransformInputIterator<bool, NonZeroOp<scalar_t>, scalar_t*> itr(self_.data_ptr<scalar_t>(), NonZeroOp<scalar_t>());
+  const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  cub::DeviceReduce::Sum(d_temp_storage, temp_storage_bytes, itr, (int*)num_nonzeros.get(), N, stream);
+  auto temp_storage = allocator.allocate(temp_storage_bytes);
+  cub::DeviceReduce::Sum(temp_storage.get(), temp_storage_bytes, itr, (int*)num_nonzeros.get(), N, stream);
+  int num_nonzeros_h;
+  AT_CUDA_CHECK(cudaMemcpyAsync(&num_nonzeros_h, num_nonzeros.get(), sizeof(int), cudaMemcpyDeviceToHost, stream));
+  //need to synchronize to make sure data is available on the host
+  AT_CUDA_CHECK(cudaStreamSynchronize(stream));
+  //expected output size is num_nonzeros x ndim
+  //if someone provides an output tensor of that size, they expect that it is in-place modified
+  //but we need physical layout to be ndim x num_nonzeros, so if the passed tensor does not align
+  //we'll need to copy produced output to the passed output.
+  bool need_to_copy = out.dim()==2 && out.sizes()[0] == num_nonzeros_h && out.sizes()[1] == self.dim() && !out.t().is_contiguous();
+  at::Tensor out_temp = need_to_copy ?
+    at::empty({self.dim(), num_nonzeros_h}, out.options().device(kCUDA)) :
```

Comment:
Strange that this is needed (seems like options should also put the tensor on the correct device?). In case of CPU out may need to validate that all inputs are on CUDA first.

Reply:
Right, was my mistake, it's not needed.